File: solutions/go/word-count/1/word_count.go
package wordcount

import "unicode"

type Frequency map[string]int

func WordCount(phrase string) Frequency {
	frequency := make(Frequency)
	currentWord := ""
	apostrophe := false
	for _, c := range phrase {
		if unicode.IsLetter(c) || unicode.IsDigit(c) {
			if apostrophe {
				currentWord += "'"
				apostrophe = false
			}
			currentWord += string(unicode.ToLower(c))
		} else if c == '\'' && !apostrophe && currentWord != "" {
			apostrophe = true
		} else {
			apostrophe = false
			if currentWord != "" {
				frequency[currentWord]++
				currentWord = ""
			}
		}
	}
	if currentWord != "" {
		frequency[currentWord]++
	}
	return frequency
}

File: solutions/haskell/word-count/1/src/WordCount.hs
{-# LANGUAGE ViewPatterns #-}

module WordCount (wordCount) where

import Control.Arrow ((>>>))
import Data.Char (isDigit, isLetter, toLower)
import Data.List (dropWhileEnd)
import qualified Data.Map.Strict as Map

wordCount :: String -> [(String, Int)]
wordCount =
    splitWords
        >>> map ((,1) . map toLower . dropWhileEnd (== '\''))
        >>> Map.fromListWith (+)
        >>> Map.assocs

splitWords :: [Char] -> [String]
splitWords s =
    case dropWhile isNonWordy s of
        [] -> []
        '\'' : xs -> splitWords xs
        x : (break isNonWordy -> (w, z)) -> (x : w) : splitWords (drop 1 z)
  where
    isNonWordy c = not $ isLetter c || isDigit c || c == '\''
